### PR TITLE
#406 update location link when it does not start with http

### PR DESCRIPTION
**Pull request by asmaNode**
**DEMO:**
https://user-images.githubusercontent.com/43316837/105866415-3faea080-5ff4-11eb-9b26-55b2b64c65e3.mp4

### CODE REVIEW

File: src/esn.calendar.libs/app/components/event/form/event-form.controller.js [hunk 1]
```diff
@@ -91,6 +92,7 @@ function CalEventFormController(
   $scope.changeBackdropZIndex = changeBackdropZIndex;
   $scope.hideEventForm = true;
   $scope.isValidURL = isValidURL;
+  $scope.updateLocationLink = updateLocationLink;
```
Comment (rezk2ll) [inline src/esn.calendar.libs/app/components/event/form/event-form.controller.js:95]: should be called somewhere in the template

**Review by rezk2ll — DISMISSED** [review]

File: src/esn.calendar.libs/app/components/event/form/event-form.controller.js (hunk 1 above)
Comment (huy-ta) [inline src/esn.calendar.libs/app/components/event/form/event-form.controller.js:95]:
```suggestion
  $scope.openLocationLink = openLocationLink;
```

File: src/esn.calendar.libs/app/components/event/form/event-form.controller.js [hunk 2]
```diff
@@ -603,6 +605,15 @@ function CalEventFormController(
...
 
+  function updateLocationLink(location) {
+    let updatedLocation = location;
+
+    if (!location.startsWith('http')) {
+      updatedLocation = `//${location}`;
+    }
+    $window.open(updatedLocation, '_blank');
+  }
+
```
Comment (huy-ta) [inline src/esn.calendar.libs/app/components/event/form/event-form.controller.js:616]:
```suggestion
  function openLocationLink(location) {
    $window.open(location.startsWith('http') ? location : `//${location}`, '_blank', 'noopener');
  }
```

Notice that I have also added the third parameter, which is `'noopener'`. This is very important to prevent [a kind of phishing attack](https://dev.to/ben/the-targetblank-vulnerability-by-example).

File: src/esn.calendar.libs/app/components/event/form/event-form.controller.spec.js [hunk 3]
```diff
@@ -2003,6 +2005,54 @@ describe('The CalEventFormController controller', function() {
...
+      beforeEach(function() {
+        openStub = sinon.stub($window, 'open');
+      });
+
+      afterEach(function() {
+        openStub.restore();
+      });
+
+      it('should add // when the location link does not start with http ', function() {
+
```
Comment (huy-ta) [inline src/esn.calendar.libs/app/components/event/form/event-form.controller.spec.js:2018]:
```suggestion
```

File: src/esn.calendar.libs/app/components/event/form/event-form.controller.spec.js [hunk 4]
```diff
@@ -2003,6 +2005,54 @@ describe('The CalEventFormController controller', function() {
...
+    describe('the updateLocationLink method', function() {
+      beforeEach(function() {
+        openStub = sinon.stub($window, 'open');
+      });
+
+      afterEach(function() {
+        openStub.restore();
+      });
+
+      it('should add // when the location link does not start with http ', function() {
```
Comment (huy-ta) [inline src/esn.calendar.libs/app/components/event/form/event-form.controller.spec.js:2017]:
```suggestion
      it('should add // when the location link does not start with http', function() {
```

File: src/esn.calendar.libs/app/components/event/form/event-form.controller.spec.js [hunk 5]
```diff
@@ -2003,6 +2005,54 @@ describe('The CalEventFormController controller', function() {
...
+            trigger: '-PT1M',
+            attendee: 'test@open-paas.org'
+          }
+        });
+        initController();
+        scope.updateLocationLink(scope.event.location);
+
+        expect(openStub).to.have.been.calledWith('//link.com', '_blank');
+      });
+      it('should not add // when the location start with http ', function() {
```
Comment (huy-ta) [inline src/esn.calendar.libs/app/components/event/form/event-form.controller.spec.js:2036]:
```suggestion

      it('should not add // when the location start with http', function() {
```

File: src/esn.calendar.libs/app/components/event/form/event-form.controller.spec.js [hunk 6]
```diff
@@ -2003,6 +2005,54 @@ describe('The CalEventFormController controller', function() {
...
+          alarm: {
+            trigger: '-PT1M',
+            attendee: 'test@open-paas.org'
+          }
+        });
+        initController();
+        scope.updateLocationLink(scope.event.location);
+        expect(openStub).to.have.been.calledWith('http://link.com', '_blank');
+      });
+
```
Comment (huy-ta) [inline src/esn.calendar.libs/app/components/event/form/event-form.controller.spec.js:2053]:
```suggestion
```

File: src/esn.calendar.libs/app/components/event/form/event-form.pug [hunk 7]
```diff
@@ -83,7 +83,7 @@ form.event-form(role="form", name="form", aria-hidden="true", ng-class="{ 'reado
                   .fg-line
                     .flex-vertical-centered.flex-start
                       input.input.location.form-control(type="text", ng-model='editedEvent.location', ng-model-options="{getterSetter: true, debounce: 500}", maxlength="{{::CAL_EVENT_FORM.location.maxlength}}", placeholder="{{ 'Location' | translate }}", ng-readonly="!canModifyEvent")
-                      a.btn.btn-link(ng-if="isValidURL(editedEvent.location)", ng-href="{{::editedEvent.location}}", target="_blank", title="{{ 'Open link' | translate }}")
+                      a.btn.btn-link(ng-if="isValidURL(editedEvent.location)", ng-click='updateLocationLink(editedEvent.location)', title="{{ 'Open link' | translate }}")
```
Comment (huy-ta) [inline src/esn.calendar.libs/app/components/event/form/event-form.pug:86]:
```suggestion
                      button.btn.btn-link(ng-if="isValidURL(editedEvent.location)", ng-click='openLocationLink(editedEvent.location)', title="{{ 'Open link' | translate }}")
```

File: src/esn.calendar.libs/app/components/event/form/event-form.controller.spec.js [hunk 8]
```diff
@@ -2003,6 +2005,54 @@ describe('The CalEventFormController controller', function() {
...
+          etag: '123456',
+          alarm: {
+            trigger: '-PT1M',
+            attendee: 'test@open-paas.org'
+          }
+        });
+        initController();
+        scope.updateLocationLink(scope.event.location);
+        expect(openStub).to.have.been.calledWith('http://link.com', '_blank');
+      });
```
Comment (huy-ta) [inline src/esn.calendar.libs/app/components/event/form/event-form.controller.spec.js:2052]:
```suggestion
        initController();

        scope.openLocationLink('http://link.com');
 
        expect(openStub).to.have.been.calledWith('http://link.com', '_blank', 'noopener');
      });
```

File: src/esn.calendar.libs/app/components/event/form/event-form.controller.js (hunk 2 above)
Comment (asmaNode) [reply src/esn.calendar.libs/app/components/event/form/event-form.controller.js]: yes i had notice that

File: src/esn.calendar.libs/app/components/event/form/event-form.controller.spec.js [hunk 9]
```diff
@@ -2165,7 +2214,6 @@ describe('The CalEventFormController controller', function() {
           });
           initController();
         });
-
```
Comment (huy-ta) [inline src/esn.calendar.libs/app/components/event/form/event-form.controller.spec.js:2168]: Why did you delete this line?

File: src/esn.calendar.libs/app/components/event/form/event-form.controller.spec.js [hunk 10]
```diff
@@ -2003,6 +2005,53 @@ describe('The CalEventFormController controller', function() {
...
+            trigger: '-PT1M',
+            attendee: 'test@open-paas.org'
+          }
+        });
+        initController();
+        scope.openLocationLink(scope.event.location);
+
+        expect(openStub).to.have.been.calledWith('//link.com', '_blank');
+      });
+      it('should not add // when the location start with http', function() {
```
Comment (huy-ta) [inline src/esn.calendar.libs/app/components/event/form/event-form.controller.spec.js:2036]:
```suggestion

      it('should not add // when the location start with http', function() {
```

**Review by huy-ta — COMMENTED** [review]
You still have the weird blank lines in your PR. Hasn't your ESLint issue been resolved already?

File: src/esn.calendar.libs/app/components/event/form/event-form.controller.spec.js (hunk 3 above)
Comment (huy-ta) [reply src/esn.calendar.libs/app/components/event/form/event-form.controller.spec.js:2018]: You haven't resolved this yet.

File: src/esn.calendar.libs/app/components/event/form/event-form.controller.spec.js (hunk 8 above)
Comment (huy-ta) [reply src/esn.calendar.libs/app/components/event/form/event-form.controller.spec.js]: You haven't resolved this.

File: src/esn.calendar.libs/app/components/event/form/event-form.controller.spec.js (hunk 10 above)
Comment (asmaNode) [reply src/esn.calendar.libs/app/components/event/form/event-form.controller.spec.js:2038]: I didn't update the ESLint config until the last time

File: src/esn.calendar.libs/app/components/event/form/event-form.controller.spec.js (hunk 8 above)
Comment (huy-ta) [reply src/esn.calendar.libs/app/components/event/form/event-form.controller.spec.js]: I mean you don't need to initialize `scope.event` to test the `openLocationLink` function. You just need to pass in a URL string.

File: src/esn.calendar.libs/app/components/event/form/event-form.controller.spec.js [hunk 11]
```diff
@@ -2003,6 +2005,58 @@ describe('The CalEventFormController controller', function() {
...
+            trigger: '-PT1M',
+            attendee: 'test@open-paas.org'
+          }
+        });
+
+        initController();
+
+        scope.openLocationLink(scope.event.location);
+
+        expect(openStub).to.have.been.calledWith('//link.com', '_blank');
```
Comment (huy-ta) [inline src/esn.calendar.libs/app/components/event/form/event-form.controller.spec.js:2035]:
```suggestion
        expect(openStub).to.have.been.calledWith('//link.com', '_blank', 'noopener');
```

File: src/esn.calendar.libs/app/components/event/form/event-form.controller.spec.js (hunk 8 above)
Comment (asmaNode) [reply src/esn.calendar.libs/app/components/event/form/event-form.controller.spec.js]: we must initialize , we got 
```should return false if the location is not valid
     The CalEventFormController controller The CalEventFormController controller the isValidURL method
     $scope.event is undefined```

**Review by rezk2ll — APPROVED** [review]

**Review by huy-ta — APPROVED** [review]
